src/v/storage/tests/storage_e2e_test.cc

```diff
@@ -2129,3 +2132,87 @@ FIXTURE_TEST(test_querying_term_last_offset, storage_test_fixture) {
 
     BOOST_REQUIRE(!log.get_term_last_offset(model::term_id(0)).has_value());
 }
+
+static storage::log_gap_analysis analyze(storage::log::impl& log) {
+    // TODO factor out common constant
+    storage::log_reader_config reader_cfg(
+      model::offset(0),
+      model::model_limits<model::offset>::max(),
+      0,
+      10_MiB,
+      ss::default_priority_class(),
+      std::nullopt,
+      std::nullopt,
+      std::nullopt);
+    return storage::make_log_gap_analysis(
+      log.make_reader(reader_cfg).get(), model::offset(0));
+}
+
+FIXTURE_TEST(test_max_compact_offset, storage_test_fixture) {
+    // Test setup.
+    auto cfg = default_log_config(test_dir);
+    cfg.max_segment_size = config::mock_binding<size_t>(10_MiB);
+    cfg.stype = storage::log_config::storage_type::disk;
+    ss::abort_source as;
+    storage::log_manager mgr = make_log_manager(cfg);
+    info("Configuration: {}", mgr.config());
+    auto deferred = ss::defer([&mgr]() mutable { mgr.stop().get0(); });
+    auto ntp = model::ntp("default", "test", 0);
+    storage::ntp_config::default_overrides overrides;
+    overrides.cleanup_policy_bitflags
+      = model::cleanup_policy_bitflags::compaction;
+    storage::ntp_config ntp_cfg(
+      ntp,
+      mgr.config().base_dir,
+      std::make_unique<storage::ntp_config::default_overrides>(overrides));
+    auto log = mgr.manage(std::move(ntp_cfg)).get0();
+    auto disk_log = get_disk_log(log);
+
+    // (1) append some random data, with limited number of distinct keys, so
```

Comment:
Can we have at least a single test with a predefined fixed set of keys? With randomness it's impossible to reproduce a failing test, also it may happen that all the keys are different and compaction happened beyond the allowed boundary but because all the keys are distinct it had zero effect and the test wouldn't fail when it should.

Before we use randomness we should be certain that the basic scenarios work:

- defined max_compactable_offset blocks compaction
- default max_compactable_offset allows compaction to happen

Reply:
For sure, happy to add a simple, deterministic case as well.

Reply:
I've added two additional test cases without the message randomization, with and without setting the max compactable offset.